### PR TITLE
FEATURE - complete block orders

### PR DESCRIPTION
**Pull request by dannypaz**
## Description
This PR adds functionality to the broker that allows us to take a block order into a `COMPLETED` status once all orders have been fulfilled.

Changes:
1. added events for completed/rejected
2. Created a new state machine plugin called `StateMachineEvents`
3. used EventEmitters for events in a order to communicate with a block order through the use of `StateMachineEvents`
4. updated block order to only set its complete status if all orders/fills are fulfilled
5. tests

## Related PRs
List related PRs if applicable

## Todos
- [ ] Tests
- [ ] Documentation
- [x] Link to Trello

### CODE REVIEW

**Comment by dannypaz** [comment]
[Change order status to COMPLETED when it's completed [CCXT]](https://trello.com/c/fBfocb4m/635-change-order-status-to-completed-when-its-completed-ccxt)

**Comment by dannypaz** [comment]
[inform clients of orders that are cancelled bc the relayer went down](https://trello.com/c/NjzCIgwD/620-inform-clients-of-orders-that-are-cancelled-bc-the-relayer-went-down)

File: broker-cli/certs/.gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
-# Ignore everything in this directory except this file
+# Ignore everything in this directory except this file and a README
```
Comment (treygriffith) [inline broker-cli/certs/.gitignore]: I think this is the wrong pr to make this change in.

File: broker-cli/certs/.gitignore [hunk 2]
```diff
@@ -1,4 +1,5 @@
...
 # We never want to inadvertantly check-in files to the certs/ directory
+*
```
Comment (treygriffith) [inline broker-cli/certs/.gitignore]: Don't make this change in this PR.

File: broker-cli/test/test-helper.js [hunk 3]
```diff
@@ -9,7 +9,10 @@
 // unhandled promise rejections are not handled by default in mocha currently,
 // but are deprecated in node.js, so should indicate test failure
 // @see {@link https://github.com/mochajs/mocha/issues/2640|mochajs/mocha#2640}
-process.on('unhandledRejection', () => { throw new Error('Unhandled rejection during testing') })
+process.on('unhandledRejection', (err) => {
+  console.error(err)
```
Comment (treygriffith) [inline broker-cli/test/test-helper.js]: I don't think this error is the one that you want. Have you tested this?

File: broker-daemon/block-order-worker/index.js [hunk 4]
```diff
@@ -1,17 +1,25 @@
...
 const { promisify } = require('util')
 const safeid = require('generate-safe-id')
+
 const { BlockOrder, Order, Fill } = require('../models')
 const { OrderStateMachine, FillStateMachine } = require('../state-machines')
+const {
+  Big,
+  getRecords,
+  SublevelIndex,
+  events: { BlockOrderWorkerEvents }
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: I don't think this should be in utils since it's specific to this module

File: broker-daemon/block-order-worker/index.js [hunk 5]
```diff
@@ -91,7 +126,26 @@ class BlockOrderWorker extends EventEmitter {
 
     this.logger.info(`Created and stored block order`, { blockOrderId: blockOrder.id })
 
-    this.emit('BlockOrder:create', blockOrder)
+    // Register listener events by the current blockOrderId
+    this.once(BlockOrderWorkerEvents.COMPLETE + id, async (blockOrderId) => this.completeBlockOrder(blockOrderId))
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: I don't like the concatenation here, I'd rather this be a function I think.

File: broker-daemon/models/block-order.js [hunk 6]
```diff
@@ -160,7 +160,24 @@ class BlockOrder {
...
+   * @return {BlockOrder} Modified block order instance
+   */
+  complete () {
+    const activeOrders = this.openOrders.filter(o => o.status !== BlockOrder.STATUSES.COMPLETED)
+    const activeFills = this.fills.filter(f => f.status !== BlockOrder.STATUSES.COMPLETED)
+
+    // If we still have active orders or fills, then the block order is not completed
+    // so we can simply return the object
+    if (activeOrders || activeFills) {
+      return this
```
Comment (treygriffith) [inline broker-daemon/models/block-order.js]: It seems odd to not error out if we aren't updating the status. Why do that?

File: broker-daemon/models/block-order.js [hunk 7]
```diff
@@ -160,7 +160,24 @@ class BlockOrder {
...
     this.status = BlockOrder.STATUSES.FAILED
+    return this
+  }
+
+  /**
+   * Move the block order to a completed status
+   * @return {BlockOrder} Modified block order instance
+   */
+  complete () {
+    const activeOrders = this.openOrders.filter(o => o.status !== BlockOrder.STATUSES.COMPLETED)
```
Comment (treygriffith) [inline broker-daemon/models/block-order.js]: Can't a block order be completed even if an underlying order was cancelled?

File: broker-daemon/models/block-order.js [hunk 8]
```diff
@@ -160,7 +160,24 @@ class BlockOrder {
...
+    const activeFills = this.fills.filter(f => f.status !== BlockOrder.STATUSES.COMPLETED)
```
Comment (treygriffith) [inline broker-daemon/models/block-order.js]: See comment above: can't a block order be completed even if a fill was rejected (shouldn't we even expect that case)?

File: broker-daemon/state-machines/fill-state-machine.js [hunk 9]
```diff
@@ -111,10 +114,11 @@ const FillStateMachine = StateMachine.factory({
    * @param  {RelayerClient}       options.relayer
    * @param  {Map<String, Engine>} options.engines     Collection of all avialable engines
    * @param  {Function}            options.onRejection A function to handle rejections of the fill
+   * @param  {Function}            options.onCompletion A function to handle the completion of the fill
    * @return {Object}                                  Data to attach to the state machine
    */
-  data: function ({ store, logger, relayer, engines, onRejection = function () {} }) {
-    return { store, logger, relayer, engines, onRejection, fill: {} }
+  data: function ({ store, logger, relayer, engines, worker }) {
```
Comment (treygriffith) [inline broker-daemon/state-machines/fill-state-machine.js]: I'm not a fan of exposing the block order worker to the state machine. I think the worker should know about the state machine, but not the inverse of that. I don't think there is a need to make that relationship two-way.

File: broker-daemon/block-order-worker/index.js (hunk 4 above)
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: it is shared between modules

File: broker-daemon/models/block-order.js (hunk 6 above)
Comment (dannypaz) [reply broker-daemon/models/block-order.js]: This is a future change. If the block order still has active orders/fills we dont want to update its status to completed

File: broker-daemon/models/block-order.js (hunk 7 above)
Comment (dannypaz) [reply broker-daemon/models/block-order.js]: Yup. im missing this case

File: broker-daemon/state-machines/fill-state-machine.js (hunk 9 above)
Comment (dannypaz) [reply broker-daemon/state-machines/fill-state-machine.js]: @treygriffith its not necessarily exposing the worker as much as exposing the events API. The current use of `onRejection` was ok, but its real goal is message passing. The only reason we use `worker` here is for triggering an event that the block order is listening for, that the block order can only set.

Comment (dannypaz) [reply broker-daemon/state-machines/fill-state-machine.js]: @treygriffith This was the tradeoff that I made vs using a callback. I tried to avoid this situation because we both had agreed earlier that we wanted the worker/state machine separation. This was one of the cleanest implementation choices

File: broker-cli/test/test-helper.js (hunk 3 above)
Comment (dannypaz) [reply broker-cli/test/test-helper.js]: @treygriffith I had legit test failures that were being swallowed by this call. We should log the error JIC

**Comment by dannypaz** [comment]
After talking w/ @treygriffith we have decided to have the state machines be event emitters, where the block order worker will add events directly on the state machine in order to pass messages between the two.

File: broker-daemon/block-order-worker/index.js (hunk 4 above)
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: @treygriffith we are removing this in the next iteration of the PR

File: broker-daemon/block-order-worker/index.spec.js [hunk 10]
```diff
@@ -1090,25 +1088,14 @@ describe('BlockOrderWorker', () => {
...
+    it('registers an event on an order for order completion', async () => {
       await worker._placeOrder(blockOrder, '100')
-
-      expect(OrderStateMachine.create).to.have.been.calledWith(sinon.match({ onRejection: sinon.match.func }))
+      expect(onceStub).to.have.been.calledWith('complete', sinon.match.func)
     })
 
-    it('fails the block order if the order is rejected', async () => {
-      worker.failBlockOrder = sinon.stub()
+    it('registers an event on an order for order rejection', async () => {
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.spec.js]: don't you need to test the function that is registered as well?

File: broker-daemon/models/block-order.js [hunk 11]
```diff
@@ -159,8 +159,26 @@ class BlockOrder {
...
+   * @return {BlockOrder} Modified block order instance
+   */
+  complete () {
+    const activeOrders = this.openOrders.filter(o => o.status === BlockOrder.STATUSES.ACTIVE)
+    const activeFills = this.fills.filter(f => f.status === BlockOrder.STATUSES.ACTIVE)
+
+    // If we still have active orders or fills, then the block order is not completed
+    // so we can simply return the object
+    if (activeOrders.length > 0 || activeFills.length > 0) {
+      return this
```
Comment (treygriffith) [inline broker-daemon/models/block-order.js]: shouldn't this be an error state of some kind?

File: broker-daemon/relayer/relayer-client.spec.js [hunk 12]
```diff
@@ -36,6 +37,11 @@ describe('RelayerClient', () => {
   let ENABLE_SSL = true
 
   beforeEach(() => {
+    logger = {
```
Comment (treygriffith) [inline broker-daemon/relayer/relayer-client.spec.js]: can this be in another pr? seems unrelated to this change.

File: broker-daemon/state-machines/plugins/state-machine-events.js [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Events that should trigger the cleanup of all event handlers created
+ * by the StateMachineEvents plugin
+ *
+ * @constant
+ * @type {Array<String>}
+ * @default
+ */
+const EXIT_EVENTS = Object.freeze(['cancel', 'complete', 'reject'])
```
Comment (treygriffith) [inline broker-daemon/state-machines/plugins/state-machine-events.js]: I think these plugins should be generic, something we would be comfortable using for any state machine.

This should be pulled into configuration for the plugin.

File: broker-daemon/state-machines/plugins/state-machine-events.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+    return {
+      /**
+       * @param {Object} lifecycle
+       * @returns {void}
+       */
+      onAfterTransition: function (lifecycle) {
+        // We check to see if an order exists, if it does then we are dealing with an
+        // OSM, or else we use the FSM
+        const blockOrderId = this.order ? this.order.blockOrderId : this.fill.blockOrderId
```
Comment (treygriffith) [inline broker-daemon/state-machines/plugins/state-machine-events.js]: the state machine should not have knowledge of the block order, and this plugin should not have knowledge of how the order and fill state machines are structured.

We shouldn't be emitting the block order id at all - the block order is itself listening to this event emitter, so the ID is superfluous. Emitting the lifecycle event is plenty specific enough.

File: broker-daemon/state-machines/plugins/state-machine-events.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  /**
+   * Alias methods on a StateMachine for event handling
+   */
+  get methods () {
+    const { eventHandler } = this
+
+    return {
+      once: (type, cb) => eventHandler.once(type, cb)
```
Comment (treygriffith) [inline broker-daemon/state-machines/plugins/state-machine-events.js]: why expose only `once`? why not provide the full API? e.g. `on`, `once`, `removeListener`, etc

File: broker-daemon/block-order-worker/index.js [hunk 16]
```diff
@@ -126,26 +114,22 @@ class BlockOrderWorker extends EventEmitter {
...
-      this.removeAllListeners(BlockOrderWorkerEvents.CREATED + id)
-    })
+    // Move BlockOrder to a rejected state if any process in working a block order, fails
+    this.once(rejectedEventId, async (blockOrderId, err) => this.failBlockOrder(blockOrderId, err))
 
-    this.on(BlockOrderWorkerEvents.CREATED + id, async (blockOrder) => {
-      await this.workBlockOrder(blockOrder).catch(e => this.emit(BlockOrderWorkerEvents.REJECTED, blockOrder.id, e))
+    // Start working the block order in another process to prevent blocking the creation
+    // of 'other' block orders
+    this.once(createEventId, async (blockOrder) => {
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: @treygriffith I don't think there is any reason that would be an issue.

File: broker-daemon/block-order-worker/index.js [hunk 17]
```diff
@@ -1,17 +1,45 @@
...
+ * Event names used for handling emitted events during the 'working' of a single
+ * block order
+ *
+ * @constant
+ * @type {Object}
+ * @default
+ */
+const WORKER_EVENTS = Object.freeze({
+  CREATE: 'CREATE',
+  REJECTED: 'REJECTED'
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: @treygriffith this was a part of the original implementation. We are going to remove the `CREATE` event

File: broker-daemon/block-order-worker/index.js [hunk 18]
```diff
@@ -60,7 +81,9 @@ class BlockOrderWorker extends EventEmitter {
   }
 
   /**
-   * Create a new block order
+   * Creates a new block order and registers events for all orders under a block order
+   *
+   * @param {Object} options
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: @treygriffith you'll need to find a source. This hasn't been the case in my experience. If we omit `@param {Object} options` then you'll have parent properties that say `options.whatever` w/ out proper nesting in the template.

File: broker-daemon/block-order-worker/index.js [hunk 19]
```diff
@@ -91,7 +114,22 @@ class BlockOrderWorker extends EventEmitter {
...
+    // Register events on a block order worker to handle processing of concurrent orders
+    const createEventId = generateIdForEvent(WORKER_EVENTS.CREATE, id)
+    const rejectedEventId = generateIdForEvent(WORKER_EVENTS.REJECTED, id)
+
+    // Move BlockOrder to a rejected state if any process in working a block order, fails
+    this.once(rejectedEventId, async (blockOrderId, err) => this.failBlockOrder(blockOrderId, err))
+
+    // Start working the block order in another process to prevent blocking the creation
+    // of 'other' block orders
+    this.once(createEventId, async (blockOrder) => {
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: Dupe comment. see reply above

File: broker-daemon/models/block-order.js (hunk 11 above)
Comment (dannypaz) [reply broker-daemon/models/block-order.js]: I made a comment above. We shouldn't complete an BlockOrder if there are existing active fills or orders. This is assuming that a BlockOrder -> has many orders/fills.

File: broker-daemon/block-order-worker/index.js [hunk 20]
```diff
@@ -345,15 +406,20 @@ class BlockOrderWorker extends EventEmitter {
...
-        }
+        store
       },
       blockOrder.id,
       { side, baseSymbol, counterSymbol, baseAmount, counterAmount }
     )
 
+    // Register listener events for the current order
+    const onceCompleteEvent = async (blockOrderId) => this.completeBlockOrder(blockOrderId)
+    const onceRejectedEvent = async (blockOrderId) => this.failBlockOrder(blockOrderId, order.error)
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: Doesnt need to be. works all the same

File: broker-daemon/block-order-worker/index.js [hunk 21]
```diff
@@ -345,15 +406,20 @@ class BlockOrderWorker extends EventEmitter {
...
-          this.failBlockOrder(blockOrder.id, err)
-        }
+        store
       },
       blockOrder.id,
       { side, baseSymbol, counterSymbol, baseAmount, counterAmount }
     )
 
+    // Register listener events for the current order
+    const onceCompleteEvent = async (blockOrderId) => this.completeBlockOrder(blockOrderId)
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: Doesnt need to be. works all the same

File: broker-daemon/block-order-worker/index.js [hunk 22]
```diff
@@ -222,12 +256,10 @@ class BlockOrderWorker extends EventEmitter {
    * @return {void}
    */
   async failBlockOrder (blockOrderId, err) {
-    // TODO: expose the stack trace of err because it is hard to troubleshoot without it
-    this.logger.error('Error encountered while working block', { id: blockOrderId, error: err.message })
+    this.logger.error('Error encountered while working block', { id: blockOrderId, error: err.stack })
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: @treygriffith there was a comment that said `// TODO: include stack information because this error is hard to debug`, so I added the full stack

File: broker-daemon/block-order-worker/index.js [hunk 23]
```diff
@@ -91,7 +114,22 @@ class BlockOrderWorker extends EventEmitter {
...
+    // Move BlockOrder to a rejected state if any process in working a block order, fails
+    this.once(rejectedEventId, async (blockOrderId, err) => this.failBlockOrder(blockOrderId, err))
+
+    // Start working the block order in another process to prevent blocking the creation
+    // of 'other' block orders
+    this.once(createEventId, async (blockOrder) => {
+      await this.workBlockOrder(blockOrder).catch(e => this.emit(rejectedEventId, blockOrder.id, e))
+    })
+
+    // TODO: Use an ID instead of the entire BlockOrder so that we can have multiple
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: I think its a good practice to pass IDs around instead of objects because the object can be mutated before the event occurs (which could lead to issues). ill update the comment

File: broker-daemon/block-order-worker/index.spec.js [hunk 24]
```diff
@@ -202,45 +203,12 @@ describe('BlockOrderWorker', () => {
       expect(Order.fromStorage).to.have.been.calledWith(fakeKey, fakeValue)
       expect(filtered).to.be.equal(false)
     })
-
-    it('works a block order when one is created', async () => {
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.spec.js]: Test was removed from this section of the spec because we no longer use events in the initialization of a BlockOrderWorker. we test for this below.

File: broker-daemon/block-order-worker/index.spec.js [hunk 25]
```diff
@@ -202,45 +203,12 @@ describe('BlockOrderWorker', () => {
...
-      worker = new BlockOrderWorker({ orderbooks, store, logger, relayer, engines })
-      worker.workBlockOrder = sinon.stub().resolves()
-
-      await delay(15)
-
-      expect(worker.workBlockOrder).to.have.been.calledOnce()
-      expect(worker.workBlockOrder).to.have.been.calledWith(fakeBlockOrder)
-    })
-
-    it('fails a block order if working it fails', async () => {
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.spec.js]: Same as the above test. This test was removed from this section of the spec because we no longer use events in the initialization of a BlockOrderWorker. we test for this below.

File: broker-daemon/block-order-worker/index.spec.js (hunk 10 above)
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.spec.js]: Yes. I will add the test. thanks for the catch

File: broker-daemon/relayer/relayer-client.spec.js (hunk 12 above)
Comment (dannypaz) [reply broker-daemon/relayer/relayer-client.spec.js]: @treygriffith im not entirely sure where this change came from, but it shouldnt exist in the diff anymore (after updating with master)

File: broker-daemon/state-machines/plugins/state-machine-events.js (hunk 15 above)
Comment (dannypaz) [reply broker-daemon/state-machines/plugins/state-machine-events.js]: @treygriffith I've omitted those other events because they are not used

File: broker-daemon/state-machines/plugins/state-machine-events.js (hunk 14 above)
Comment (dannypaz) [reply broker-daemon/state-machines/plugins/state-machine-events.js]: @treygriffith I totally agree. I've removed the need for an ID during event handling. When these events are registered with a state machine, the caller should be providing the information needed. 

This code no longer exists

File: broker-daemon/block-order-worker/index.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: FWIW `err.stack` includes the message

File: broker-daemon/state-machines/plugins/state-machine-events.js (hunk 15 above)
Comment (treygriffith) [reply broker-daemon/state-machines/plugins/state-machine-events.js]: Ok, let's punt for now

File: broker-daemon/block-order-worker/index.js [hunk 26]
```diff
@@ -316,6 +312,37 @@ class BlockOrderWorker extends EventEmitter {
...
+  /**
+   * Move a block order to a completed state if all orders have been completed
+   * @param {String} blockOrderId
+   */
+  async completeBlockOrder (blockOrderId) {
+    this.logger.info('Attempting to put block order in a completed state', { id: blockOrderId })
+
+    // TODO: move status to its own sublevel so it can be updated atomically
+    try {
+      var value = await promisify(this.store.get)(blockOrderId)
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: since we do this twice, we should consider moving it to its own method

File: broker-daemon/block-order-worker/index.js [hunk 27]
```diff
@@ -316,6 +312,37 @@ class BlockOrderWorker extends EventEmitter {
...
+      }
+
+      throw e
+    }
+
+    const blockOrder = BlockOrder.fromStorage(blockOrderId, value)
+
+    // An order is only completed if all orders underneath the blockorder are out of
+    // an `ACTIVE` state.
+    blockOrder.complete()
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: per our discussion, any validity checks should happen here (not in the model).

Also, we should check the actual amount of the block order and not just the underlying orders/fills.

File: broker-daemon/state-machines/plugins/state-machine-events.js [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+       * @param {Object} lifecycle
+       * @returns {void}
+       */
+      onAfterTransition: function (lifecycle) {
+        plugin.eventHandler.emit(lifecycle.transition)
+
+        // Remove all events from the event emitter if we have either cancelled,
+        // completed, or rejected an order
+        if (plugin.exitEvents.includes(lifecycle.transition)) {
+          plugin.eventHandler.removeAllListeners()
```
Comment (treygriffith) [inline broker-daemon/state-machines/plugins/state-machine-events.js]: this seems strange. Why do we need this? And why do it this way?

Comment (dannypaz) [reply broker-daemon/state-machines/plugins/state-machine-events.js]: @treygriffith just a cleanup of events after we mark a StateMachine as "finished". Would hopefully free up some memory or there would be a leak

File: broker-daemon/block-order-worker/index.js [hunk 29]
```diff
@@ -316,6 +281,57 @@ class BlockOrderWorker extends EventEmitter {
...
+
+    const blockOrder = await this.getBlockOrder(blockOrderId)
+
+    // check the fillamount on each collection of state machines from the block order
+    // and make sure that either is equal to how much we are trying to fill.
+    let totalFilled = Big(0)
+
+    if (blockOrder.fills.length) {
+      totalFilled = blockOrder.fills.reduce((acc, fsm) => acc.plus(fsm.fill.fillAmount), Big(0))
+    } else if (blockOrder.openOrders.length) {
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: @treygriffith what is the situation where a single blockorder would have both of these?

File: broker-daemon/block-order-worker/index.js [hunk 30]
```diff
@@ -316,6 +281,57 @@ class BlockOrderWorker extends EventEmitter {
...
+
+    if (!activeOrders && !activeFills && !stillBeingFilled) {
+      // An order is only completed if all orders underneath the blockorder are out of
+      // an `ACTIVE` state and it is entirely filled
+      blockOrder.complete()
+      await promisify(this.store.put)(blockOrder.key, blockOrder.value)
+      this.logger.info('Moved block order to completed state', { blockOrderId })
+    } else if (stillBeingFilled && !activeFills && !activeOrders) {
+      this.logger.error('Block order is still being filled, but has no active orders/fills', { blockOrderId })
+      throw new Error('Block order in a weird state. Order is filled, but no active orders/fills still exist')
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: @treygriffith thats why we throw the error. If its in a weird state, we want to know and need to fix the bug. I dont think we should avoid this state because of a race-condition. We can simply re-run this method and the order will effectively complete, but it is still nice to know if this situation occurs.

Either way, if I should remove the check, that is fine too.

File: broker-daemon/block-order-worker/index.js [hunk 31]
```diff
@@ -316,6 +281,57 @@ class BlockOrderWorker extends EventEmitter {
...
+    } else if (!stillBeingFilled && (activeFills || activeOrders)) {
+      this.logger.error('Block order is not being filled and has active orders/fills', { activeFills, activeOrders, blockOrderId })
+      throw new Error('Block order in a weird state. Order is not being filled, but active orders/fills still exist')
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: @treygriffith I think this case is less of an issue than the one above. If the order is not being filled (meaning the fillAmounts match), then we would have effectively completed all orders (the property `active` would HAVE to be false). If that property is never switched, then we have a bug that we need to be aware of. I dont see an edgecase for this piece of code.

File: broker-daemon/state-machines/order-state-machine.js [hunk 32]
```diff
@@ -362,15 +383,7 @@ const OrderStateMachine = StateMachine.factory({
...
-    },
-
-    /**
-     * Handle rejected state by calling a passed in handler
-     * @param  {Object} lifecycle Lifecycle object passed by javascript-state-machine
-     * @return {void}
-     */
-    onAfterReject: async function (lifecycle) {
-      this.onRejection(this.error)
+      this.order.error = error
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/state-machines/order-state-machine.js]: @treygriffith yeah, we lost it with the removal of the method call

File: broker-daemon/block-order-worker/index.js (hunk 30 above)
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: @treygriffith I've had a second thought. If this piece of code makes the `complete` method too complicated (or weird) then we should remove it. If this issue occurs, it is probably better to realize and error at a different point in the lifecycle, than at the very end. What do you think?

File: broker-daemon/models/order.js [hunk 33]
```diff
@@ -246,6 +246,16 @@ class Order {
...
   }
 
+  /**
+   * @return {Boolean}
+   */
+  get active () {
+    // We currently define an "ACTIVE" order as an Order w/out a fillAmount, since this
+    // property is only set once an order has been filled.
+    // TODO: Figure out a better definition of an "ACTIVE" order (setting a status?)
+    return (!this.swapPreimage)
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/models/order.js]: ty good catch @treygriffith

File: broker-daemon/block-order-worker/index.js [hunk 34]
```diff
@@ -397,21 +419,28 @@ class BlockOrderWorker extends EventEmitter {
...
       )
+
+      // The events below tie into the StateMachine's lifecycle hooks for a OrderStateMachine
+      fsm.once('execute', () => {
+        this.completeBlockOrder(blockOrder.id).catch(e => {
+          this.logger.error(`BlockOrder failed to be completed, status is left active: ${blockOrder.id}`, { error: e.stack })
+        })
+      })
+
+      fsm.once('reject', () => this.failBlockOrder(blockOrder.id, fsm.fill.error))
```
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: Totally agree. At the very least it'd be nice to log that we failed to fail

File: broker-daemon/block-order-worker/index.js (hunk 29 above)
[the remark this answers is not in the document]
Comment (treygriffith) [reply broker-daemon/block-order-worker/index.js]: limit order where there are existing orders on the books at that price

File: broker-daemon/block-order-worker/index.js [hunk 35]
```diff
@@ -316,6 +281,43 @@ class BlockOrderWorker extends EventEmitter {
...
+      return acc.plus(osm.order.fillAmount || 0)
+    }, totalFilled)
+
+    this.logger.debug('Current total filled amount: ', { totalFilled, blockOrderAmount: blockOrder.baseAmount })
+
+    const stillBeingFilled = totalFilled.lt(blockOrder.baseAmount)
+    const activeOrders = blockOrder.openOrders.some(osm => osm.order.active)
+    const activeFills = blockOrder.fills.some(fsm => fsm.fill.active)
+
+    if (!activeOrders && !activeFills && !stillBeingFilled) {
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: if it is not still being filled and it has active orders, what behavior do we want?

File: broker-daemon/block-order-worker/index.js [hunk 36]
```diff
@@ -316,6 +281,43 @@ class BlockOrderWorker extends EventEmitter {
...
+      // An order is only completed if all orders underneath the blockorder are out of
+      // an `ACTIVE` state and it is entirely filled
+      blockOrder.complete()
+      await promisify(this.store.put)(blockOrder.key, blockOrder.value)
+      this.logger.info('Moved block order to completed state', { blockOrderId })
+    } else {
+      this.logger.debug('Block order is not ready to be completed', { blockOrderId })
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: should this be an error?

File: broker-daemon/block-order-worker/index.js [hunk 37]
```diff
@@ -340,21 +342,38 @@ class BlockOrderWorker extends EventEmitter {
...
 
-    this.logger.info('Created order for BlockOrder', { blockOrderId: blockOrder.id, orderId: order.orderId })
+    // We are hooking into the complete lifecycle event of an order state machine to trigger
+    // the completion of a blockorder
+    osm.once('complete', async () => {
+      this.completeBlockOrder(blockOrder.id)
+        .catch(e => {
+          this.logger.error(`BlockOrder failed to be completed from order`, { id: blockOrder.id, error: e.stack })
+        })
+        .then(() => osm.removeAllListeners())
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: if we are using a `once`, do we need to clean up listeners here?

File: broker-daemon/models/fill.js [hunk 38]
```diff
@@ -243,6 +243,16 @@ class Fill {
...
   }
 
+  /**
+   * @return {Boolean}
+   */
+  get active () {
+    // We define an "ACTIVE" fill as a Fill that has not been assigned a maker address
+    // meaning that the fill has no been set to execute.
+    // TODO: Figure out a better definition of an "ACTIVE" fill (setting a status?)
+    return (!this.makerAddress)
```
[the remark this answers is not in the document]
Comment (treygriffith) [reply broker-daemon/models/fill.js]: Adding on to this, it may make sense to make sure the fill hasn't been rejected

File: broker-daemon/models/order.js (hunk 33 above)
[the remark this answers is not in the document]
Comment (treygriffith) [reply broker-daemon/models/order.js]: This still needs to be fixed @dannypaz

File: broker-daemon/block-order-worker/index.spec.js (hunk 10 above)
Comment (treygriffith) [reply broker-daemon/block-order-worker/index.spec.js]: @dannypaz where is the test for the registered functions?

File: broker-daemon/block-order-worker/index.js (hunk 35 above)
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: @treygriffith we intentionally removed checks for that functionality

File: broker-daemon/block-order-worker/index.js (hunk 36 above)
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: Not necessarily. There are cases where this would be an error, however we decided to remove them. As an example: This would be hit if the block order is still being filled, which is a valid case and wouldn't be an error

File: broker-daemon/block-order-worker/index.js (hunk 37 above)
Comment (dannypaz) [reply broker-daemon/block-order-worker/index.js]: We still need to clean up, because we would have a `reject` listener still open on the `osm`

File: broker-daemon/models/order.js (hunk 33 above)
[the remark this answers is not in the document]
Comment (dannypaz) [reply broker-daemon/models/order.js]: @treygriffith we decided to punt this. there is a ticket in the backlog for adding a `status` to an order.

File: broker-daemon/block-order-worker/index.js [hunk 39]
```diff
@@ -340,21 +342,40 @@ class BlockOrderWorker extends EventEmitter {
...
+
+    // We are hooking into the reject lifecycle event of an order state machine to trigger
+    // the failure of a blockorder
+    osm.once('reject', async () => {
+      this.failBlockOrder(blockOrder.id, osm.order.error)
+        .catch(e => {
+          this.logger.error(`BlockOrder failed on setting a failed status from order`, { id: blockOrder.id, error: e.stack })
+        })
+        .then(() => {
+          return osm.removeAllListeners()
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: [nit] weird that you did multiple lines on this one but single line for `complete`

File: broker-daemon/models/block-order.js [hunk 40]
```diff
@@ -263,6 +276,30 @@ class BlockOrder {
...
+   * Grab a block order from a given sublevel
+   *
+   * @param {Sublevel} store block order sublevel store
+   * @param {String} blockOrderId
+   * @return {BlockOrder} BlockOrder instance
+   * @throws {Error} store is null
+   * @throws {BlockOrderNotFoundError} block order could not be found
+   */
+  static async fromStore (store, blockOrderId) {
+    if (!store) throw new Error('[BlockOrder#fromStore] No leveldb store is defined')
```
Comment (treygriffith) [inline broker-daemon/models/block-order.js]: [nit] I don't think there is a good reason to reference leveldb in the error message

File: broker-daemon/models/order.js [hunk 41]
```diff
@@ -246,6 +246,16 @@ class Order {
     }
   }
 
+  /**
+   * @return {Boolean}
+   */
+  get active () {
```
Comment (treygriffith) [inline broker-daemon/models/order.js]: why define this in this PR if we aren't using it?

File: broker-daemon/block-order-worker/index.spec.js [hunk 42]
```diff
@@ -885,14 +794,140 @@ describe('BlockOrderWorker', () => {
...
+            { order: { fillAmount: 100 } }
+          ]
+        }
+        getBlockOrderStub.returns(activeBlockOrder)
+        await worker.checkBlockOrderCompletion(blockOrder.id)
+        expect(blockOrderCompleteStub).to.not.have.been.calledOnce()
+      })
+    })
+
+    context('block order with fills', () => {
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.spec.js]: consider adding a test with a block order with both orders and fills

File: broker-daemon/block-order-worker/index.spec.js [hunk 43]
```diff
@@ -885,14 +794,140 @@ describe('BlockOrderWorker', () => {
...
+        worker.blockOrder = blockOrder
+        worker.getBlockOrder = getBlockOrderStub.returns(blockOrder)
+      })
+
+      it('completes a block order', async () => {
+        await worker.checkBlockOrderCompletion(blockOrder.id)
+        expect(blockOrderCompleteStub).to.have.been.calledOnce()
+      })
+
+      it('does not complete a block order if order is still active', async () => {
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.spec.js]: is this test mis-named? the underlying orders being active is no longer a condition

File: broker-daemon/models/order.js (hunk 41 above)
Comment (dannypaz) [reply broker-daemon/models/order.js]: Good catch. I forgot to remove these from the order/fill.
